### ISSUE TITLE
Erroneous compilation error

### ISSUE THREAD

**Issue by paganaye** [Microsoft/TypeScript#14318]
**TypeScript Version:**  2.2.1

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
class ClassA {
   value: number = 0;

   setValue(newValue: number) {
      this.value = newValue;
   }

   test() {
      if (this.value !== 0) throw 'Error';
      this.setValue(1);
      if (this.value == 1) alert('Value has changed')
   }
}

var a = new ClassA();

a.test();
```

**Expected behavior:**

The test is valid, it evens returns true.

**Actual behavior:**

The typescript compiler is showing a compilation error:
file: 'file:///home/pascal/Code/eval/test.ts'
severity: 'Error'
message: 'Operator '==' cannot be applied to types '0' and '1'.'
at: '11,11'
      if (this.value == 1) alert('Value has changed')

**Comment by paganaye** [comment]
One way to circumvent the issue is to cast the value:

```ts
     if (this.value as number == 1) alert('Value has changed')
```

**Comment by DanielRosenwasser** [comment]
This is one of the few places that control flow analysis can be frustrating. We've had discussions about it in the past. Check out #9998 for an in-depth discussion on this.
